File: python/입국심사.py
def binarysearch(n, answer_list, s, e, t):

    if s == e or s+2 == e:
        return max(answer_list)
    w = (s+e)//2
    sum = 0
    temp = 0
    for i in range(len(t)):
        temp = w // t[i]
        sum += temp
        answer_list[i] = temp * t[i]
    if sum > n:
        sum = binarysearch(n, answer_list, s, w, t)
    elif sum < n:
        sum = binarysearch(n, answer_list, w, e, t)
    return max(answer_list)

def solution(n, times):
    answer = 0
    times = sorted(times)
    answer_list = [0] * len(times)
    answer = binarysearch(n, answer_list, 0, max(times)*n, times)
    
    return answer

print(solution(2, [1, 2]))
#print(solution(3, [1, 2, 3]))

File: JavaScript/입국심사.js
function solution(n, times) {
    if (times.length === 1) return n * times[0];
    times.sort((a, b) => { return a - b });
    let min = 0, max = Math.max.apply(null, times) * n;
    let answerList = new Array(times.length);
    while (min !== max) {
        const ave = Math.floor((min + max) / 2);
        let totalPeople = 0;
        for (let i = 0; i < times.length; i++) {
            const p = Math.floor(ave / times[i]);
            totalPeople += p;
            answerList[i] = p * times[i];
        }
        if (min === ave) {
            for (let j = 0; j < times.length; j++) {
                answerList[j] += times[j];
            }
            return Math.min.apply(null, answerList);
        }
        if (n < totalPeople) {
            max = ave;
        } else if (n > totalPeople) {
            min = ave;
        } else {
            break;
        }
    }
    return Math.max.apply(null, answerList);
}
